samples/net/aws_iot/src/main.c

```diff
@@ -0,0 +1,402 @@
+	case AWS_IOT_EVT_CONNECTING:
+		LOG_INF("AWS_IOT_EVT_CONNECTING");
+		break;
+	case AWS_IOT_EVT_CONNECTED:
+		LOG_INF("AWS_IOT_EVT_CONNECTED");
+		on_aws_iot_evt_connected(evt);
+		break;
+	case AWS_IOT_EVT_READY:
+		LOG_INF("AWS_IOT_EVT_READY");
+		break;
+	case AWS_IOT_EVT_DISCONNECTED:
+		LOG_INF("AWS_IOT_EVT_DISCONNECTED");
+		on_aws_iot_evt_disconnected();
+		break;
+	case AWS_IOT_EVT_DATA_RECEIVED:
+		LOG_INF("AWS_IOT_EVT_DATA_RECEIVED");
+
+		LOG_INF("Received message: \"%.*s\" on topic: \"%.*s\"", evt->data.msg.len,
+									 evt->data.msg.ptr,
+									 evt->data.msg.topic.len,
+									 evt->data.msg.topic.str);
+		break;
+	case AWS_IOT_EVT_PUBACK:
+		LOG_INF("AWS_IOT_EVT_PUBACK, message ID: %d", evt->data.message_id);
+		break;
+	case AWS_IOT_EVT_PINGRESP:
+		LOG_INF("AWS_IOT_EVT_PINGRESP");
+		break;
+	case AWS_IOT_EVT_FOTA_START:
+		LOG_INF("AWS_IOT_EVT_FOTA_START");
+		break;
+	case AWS_IOT_EVT_FOTA_ERASE_PENDING:
+		LOG_INF("AWS_IOT_EVT_FOTA_ERASE_PENDING");
+		break;
+	case AWS_IOT_EVT_FOTA_ERASE_DONE:
+		LOG_INF("AWS_FOTA_EVT_ERASE_DONE");
+		break;
+	case AWS_IOT_EVT_FOTA_DONE:
+		LOG_INF("AWS_IOT_EVT_FOTA_DONE");
+		on_aws_iot_evt_fota_done(evt);
+		break;
+	case AWS_IOT_EVT_FOTA_DL_PROGRESS:
+		LOG_INF("AWS_IOT_EVT_FOTA_DL_PROGRESS, (%d%%)", evt->data.fota_progress);
+		break;
+	case AWS_IOT_EVT_ERROR:
+		LOG_INF("AWS_IOT_EVT_ERROR, %d", evt->data.err);
+		break;
+	case AWS_IOT_EVT_FOTA_ERROR:
+		LOG_INF("AWS_IOT_EVT_FOTA_ERROR");
+		break;
+	default:
+		LOG_WRN("Unknown AWS IoT event type: %d", evt->type);
+		break;
+	}
+}
+
+static void l4_event_handler(struct net_mgmt_event_callback *cb,
+			     uint32_t event,
+			     struct net_if *iface)
+{
+	switch (event) {
+	case NET_EVENT_L4_CONNECTED:
+		LOG_INF("IP Up");
+		on_net_event_l4_connected();
+		break;
+	case NET_EVENT_L4_DISCONNECTED:
+		LOG_INF("IP down");
+		on_net_event_l4_disconnected();
+		break;
+	default:
+		/* Don't care */
+		return;
+	}
+}
+
+static void connectivity_event_handler(struct net_mgmt_event_callback *cb,
+				       uint32_t event,
+				       struct net_if *iface)
+{
+	if (event == NET_EVENT_CONN_IF_FATAL_ERROR) {
+		LOG_ERR("NET_EVENT_CONN_IF_FATAL_ERROR");
+		FATAL_ERROR();
+		return;
+	}
+}
+
+int main(void)
+{
+	LOG_INF("The AWS IoT sample started, version: %s", CONFIG_AWS_IOT_SAMPLE_APP_VERSION);
+
+	int err;
+
+	/* Setup handler for Zephyr NET Connection Manager events. */
+	net_mgmt_init_event_callback(&l4_cb, l4_event_handler, L4_EVENT_MASK);
+	net_mgmt_add_event_callback(&l4_cb);
+
+	/* Setup handler for Zephyr NET Connection Manager Connectivity layer. */
+	net_mgmt_init_event_callback(&conn_cb, connectivity_event_handler, CONN_LAYER_EVENT_MASK);
+	net_mgmt_add_event_callback(&conn_cb);
+
+	/* Connecting to the configured connectivity layer.
+	 * Wi-Fi or LTE depending on the board that the sample was built for.
+	 */
+	LOG_INF("Bringing network interface up and connecting to the network");
+
+	err = conn_mgr_all_if_up(true);
```

Comment:
I think you can use autoconnect/auto-up now that the Wi-Fi integration has been updated to block init until WPA_supplicant is ready

Just register the callbacks at boot time rather than in `main()`

Reply:
For the same reasons mentioned in previous discussions, I want the application to decide when it wants to take the interface up or not.